### PR TITLE
ACCUMULO-4157 Bug fix for removing WALs to quickly

### PR DESCRIPTION
**Pull request by mjwall**
Keep track of first time a tserver is seen down and only remove WALs for that server if past configurated threshhold

Much work left to be done here, but trying to keep the changes to the GarbageCollectWriteAheadLogs class small to get the bug fixed.  I'll create another ticket to refactor and cleanup.

Includes an end to end test calling the collect method simulating a dead tserver.

### CODE REVIEW

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java [hunk 1]
```diff
@@ -170,76 +175,117 @@ boolean holdsLock(HostAndPort addr) {
...
+      removeSortedWAL(swalog);
+    }
+    return 0;
+  }
+
+  @VisibleForTesting
+  void removeSortedWAL(Path swalog) {
+    log.debug("Removing sorted WAL " + swalog);
+    try {
+      if (!useTrash || !fs.moveToTrash(swalog)) {
```
Comment (joshelser) [inline server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java]: Maybe lift this into a method? The general "move to trash or delete" is repeated a couple of places.

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java [hunk 2]
```diff
@@ -55,12 +56,16 @@
...
 import com.google.common.net.HostAndPort;
+import java.util.concurrent.TimeUnit;
+import org.apache.accumulo.core.conf.Property;
 
 public class GarbageCollectWriteAheadLogs {
   private static final Logger log = Logger.getLogger(GarbageCollectWriteAheadLogs.class);
 
   private final Instance instance;
   private final VolumeManager fs;
+  private final Map<HostAndPort,Long> firstSeenDead = new HashMap<HostAndPort,Long>();
```
Comment (joshelser) [inline server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java]: The safety of this not being a concurrent/synchronized map is likely due to the use of GCWALogs by the SimpleGarbageCollector? Maybe we just need to document that this class is not threadsafe now?

**Comment by joshelser** [comment]
The general approach seems to be pretty solid. The edge cases I could think about were correctly handled (e.g. if the GC dies, the "time to wait" would be reset and we woudn't delete the WAL too quickly). The cleanup of the GCWAL class is nice as are the new test cases.

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java (hunk 2 above)
Comment (mjwall) [inline server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java]: Are you suggesting I use a ConcurrentHashMap?  I thought SimpleGarbageCollector was designed to be one thread out of the [run](https://github.com/apache/accumulo/blob/1.6/server/gc/src/main/java/org/apache/accumulo/gc/SimpleGarbageCollector.java#L513) method.

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java (hunk 1 above)
Comment (mjwall) [inline server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java]: I'll add a note to the follow on ticket to really cleanup this class if that is ok?

**Comment by mjwall** [comment]
Thanks for reviewing @joshelser

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java (hunk 2 above)
Comment (joshelser) [inline server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java]: No, I don't actually think it's necessary (I haven't looked closely enough), but, at first glance, it looks like it be necessary. I believe the caller of GCWALs does so in such a way that it does not have any race conditions.

Re-worded, I think this is probably fine as-is, but some documentation saying "This is not threadsafe, but the caller is not invoking it in a concurrent manner" would be good for future devs/readers.

Comment (mjwall) [inline server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java]: Ok, great.  I'll add a javadoc to the collect method, the only public method in the class currently.

Comment (mjwall) [inline server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java]: Javadocs added

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java [hunk 3]
```diff
@@ -170,76 +184,203 @@ boolean holdsLock(HostAndPort addr) {
...
+   *
+   * @param entry
+   *          WAL information gathered
+   * @param conf
+   *          AccumuloConfiguration object
+   * @param status
+   *          GCStatus object
+   */
+  @VisibleForTesting
+  void removeWALFile(Entry<String,ArrayList<Path>> entry, AccumuloConfiguration conf, final GCStatus status) {
```
Comment (keith-turner) [inline server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java]: This method is marked as visible for testing, but I don't see the unit test calling it.

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java [hunk 4]
```diff
@@ -170,76 +184,203 @@ boolean holdsLock(HostAndPort addr) {
...
+   *          GCStatus object
+   */
+  @VisibleForTesting
+  void askTserverToRemoveWAL(HostAndPort address, AccumuloConfiguration conf, Entry<String,ArrayList<Path>> entry, final GCStatus status) {
+    firstSeenDead.remove(address);
+    Client tserver = null;
+    try {
+      tserver = ThriftUtil.getClient(new TabletClientService.Client.Factory(), address, conf);
+      tserver.removeLogs(Tracer.traceInfo(), SystemCredentials.get().toThrift(instance), paths2strings(entry.getValue()));
+      log.debug("asking tserver to delete " + entry.getValue() + " from " + entry.getKey());
```
Comment (keith-turner) [inline server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java]: I think it should say `asked` instead of `asking` because the log message is after the thrift call.

File: server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogsTest.java [hunk 5]
```diff
@@ -234,4 +257,288 @@ public void testIsUUID() {
...
+    long wait = AccumuloConfiguration.getTimeInMillis("1h");
+    long t1, t2;
+    boolean ttd;
+    do {
+      t1 = System.nanoTime();
+      gcwal.clearFirstSeenDead();
+      gcwal.timeToDelete(address, wait);
+      ttd = gcwal.timeToDelete(address, wait);
+      t2 = System.nanoTime();
+    } while (t2 - t1 > (wait / 2) * 1000000); // as long as it took less than half of the configured wait
```
Comment (keith-turner) [inline server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogsTest.java]: I saw you used TimeUnit earlier in the patch to convert time.  Could use TimeUnit earlier in this test method to convert `wait` to nanos.  IMO would be slightly more readable than multiplying by 1000000.

File: server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogsTest.java [hunk 6]
```diff
@@ -234,4 +257,288 @@ public void testIsUUID() {
...
+
+    assertFalse(gcwal.timeToDelete(address, wait));
```
Comment (keith-turner) [inline server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogsTest.java]: should this be `assertFalse(ttd)`?

File: server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogsTest.java [hunk 7]
```diff
@@ -234,4 +257,288 @@ public void testIsUUID() {
...
+    long getGCWALDeadServerWaitTime(AccumuloConfiguration conf) {
+      // tries to use zookeeper
+      return 1000l;
+    }
+  }
+
+  @Test
+  public void testCollectWithDeadTserver() throws IOException, InterruptedException {
+    Instance i = new MockInstance();
+    File walDir = new File(System.getProperty("java.io.tmpdir") + File.separator + GCWAL_DEAD_DIR);
```
Comment (keith-turner) [inline server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogsTest.java]: Most Accumulo test that create files put them in `target`.  Can use something like this do that. `System.getProperty("user.dir") + "/target/"+GCWAL_DEAD_DIR`.  The prop `user.dir` is the CWD.  Maven sets the CWD to the dir where the pom is.  Its nice having the test write to target on jenkins and environments like that.  Although maybe jenkins sets java.io.tmpdir to something reasonable.

File: server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogsTest.java [hunk 8]
```diff
@@ -234,4 +257,288 @@ public void testIsUUID() {
...
+      GCStatus status = new GCStatus(new GcCycleStats(), new GcCycleStats(), new GcCycleStats(), new GcCycleStats());
+
+      gcwal2.collect(status);
+
+      assertTrue("File should not be deleted", walFile.exists());
+      assertEquals("Should have one candidate", 1, status.lastLog.getCandidates());
+      assertEquals("Should not have deleted that file", 0, status.lastLog.getDeleted());
+
+      sleep(2000);
+      gcwal2.collect(status);
```
Comment (keith-turner) [inline server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogsTest.java]: nice test

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java [hunk 9]
```diff
@@ -410,4 +556,41 @@ static boolean isUUID(String name) {
...
+   * @param wait
+   *          long value of elapsed time to wait
+   * @return boolean whether enough time elapsed since the server was first seen as dead.
+   */
+  @VisibleForTesting
+  protected boolean timeToDelete(HostAndPort address, long wait) {
+    // check whether the tserver has been dead long enough
+    Long firstSeen = firstSeenDead.get(address);
+    if (firstSeen != null) {
+      long elapsedTime = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - firstSeen);
```
Comment (madrob) [inline server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java]: Log message claims seconds but code prints millis.

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java [hunk 10]
```diff
@@ -410,4 +556,41 @@ static boolean isUUID(String name) {
...
 
+  /**
+   * Determine if TServer has been dead long enough to remove associated WALs.
+   * <p>
+   * Uses a map where the key is the address and the value is the time first seen dead. If the address is not in the map, it is added with the current system
+   * nanoTime. When the passed in wait time has elapsed, this method returns true and removes the key and value from the map.
+   *
+   * @param address
+   *          HostAndPort of dead tserver
+   * @param wait
```
Comment (madrob) [inline server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java]: Java doc units

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java (hunk 3 above)
Comment (mjwall) [inline server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java]: Yep, good catch.  I'll remove the annotation

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java (hunk 4 above)
Comment (mjwall) [inline server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java]: Yep, will change

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java (hunk 10 above)
Comment (mjwall) [inline server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java]: Will add

File: server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java (hunk 9 above)
Comment (mjwall) [inline server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogs.java]: Good catch, I change it and didn't update the log message.  Will fix.

File: server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogsTest.java (hunk 6 above)
Comment (mjwall) [inline server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogsTest.java]: It should, good catch.

File: server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogsTest.java (hunk 5 above)
Comment (mjwall) [inline server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogsTest.java]: yep, will change

File: server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogsTest.java (hunk 7 above)
Comment (mjwall) [inline server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogsTest.java]: will update, thanks

File: server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogsTest.java [hunk 11]
```diff
@@ -234,4 +257,288 @@ public void testIsUUID() {
...
+  @Test
+  public void testTimeToDeleteFalse() {
+    HostAndPort address = HostAndPort.fromString("tserver1:9998");
+    long wait = AccumuloConfiguration.getTimeInMillis("1h");
+    long t1, t2;
+    boolean ttd;
+    do {
+      t1 = System.nanoTime();
+      gcwal.clearFirstSeenDead();
+      gcwal.timeToDelete(address, wait);
```
Comment (keith-turner) [inline server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogsTest.java]: should this first call to timeToDelete always return false?  If so could wrap it w/ `assertFalse()`

**Comment by keith-turner** [comment]
@mjwall I am finished looking at this +1

File: server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogsTest.java (hunk 11 above)
Comment (mjwall) [inline server/gc/src/test/java/org/apache/accumulo/gc/GarbageCollectWriteAheadLogsTest.java]: Added, rebased on 1.6 and and force pushed my branch.  I am ready to merge this

**Comment by mjwall** [comment]
Thanks @joshelser @keith-turner and @madrob for looking at this.  I am ready to merge

**Comment by joshelser** [comment]
>  I am ready to merge

:+1: Great detective work here!
